README.md
# **</in>Tic Tac & Toe</ins>**

[![](\_documentation/Screenshots/\_\_\_"AmiResponsive: <https://ui.dev/amiresponsive?url=>")](https://ui.dev/amiresponsive?url=/ "Visit the Proof of Responsiveness: 3 Grains & Toe")

> Visit: [3 Grains & Toe](https://ipoetdev.github.io/terni-lapilli--toe/ "3 Grains & Toe: <https://ipoetdev.github.io/terni-lapilli--toe/> by Charles J Fowler (@ipoetdev)") | [https://ipoetdev.github.io/terni-lapilli--toe/](https://ipoetdev.github.io/terni-lapilli--toe/)

[![GitHub Super-Linter](https://github.com/ipoetdev/terni-lapilli--toe/workflows/Lint%20Code%20Base/badge.svg)](https://github.com/marketplace/actions/super-linter)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

---

---

## **1. GOALS/PROPOSAL**

### _<ins>1.1 Game Rules</ins>_

Here are the basic rules of Tic Tac Toe:

- The game is played on a `3x3` grid.
- There are two players, `X` and `O`.
- Players take turns placing their symbol (`X` or `O`) on an empty cell.
- The first player to get three of their symbols (`x3`) in a row (horizontally, vertically, or diagonally) wins the game.
- If all cells on the board are occupied and no player has won, the game is a tie.

### _<ins>1.1 Developer Goals</ins>_

<!--
 Note: This developer learnt UML and software design by diagrams 20 years ago, under his Masters of Technology at RMIT (CS). This project is a showcase for some of these techniques under the Design section below.
-->

- Develop a working MVP version for a game of TicTacToe
- Develop a pattern/model of a codebase that is:
  - Modular and extensibile.
  - Configurable
  - Well documented.
  - Ease of:
    - Readability and analysability.
    - Implementation.
    - Deployment.
  - Well documented usinhg JSDoc-strings
  - Changable: Modular, uses OOP.
  - Stable, and reasonable error control flow.
  - Testible: OOP, Error Handling, Debug
  - Use of GitHub Issues to document and track errorts
- To showcase via this README, not for brevity and concisness, but showcase of my prior art and knowledge of software development (OOP, Technical Documentation, Tooling); given a Masters of Technology [2002-04] qualification.

### _<ins>1.3 User | Audience Goals</ins>_
>
> - DF: Develivered Functionality
> - FE: Future Enhancement

- [DF001](#df001) Play the board game, Tic Tac Toe: Two Player, Two Pieces: (X & O)
  - As a user wants to play a TicTacToe game, so that users can have fun and pass the time.
- [DF002](#df002) Via a simple game of Tic Tac Toe, on a single HTML, via a simple screen, i.e. a Simple Click game logic.
- [DF003](#df003) Implement defensive programming, error control, developer  friendly debugging and user feedback
  - [DF003.01](#df003.01) Use graceful error handling for User & developer alike
  - [DF003.01](#df003.02) Employ configurable debugging functionality as a debugging class for developer experience
  - [DF003.01](#df003.01) Employ simple alterts for error and app level feedback.
- [FE001](#fe001) User(s) choose game pieces, and customise the type of pieces in the process.
- [FE002](#fe002) Game Statistics to keep count of wins, looses, draws for players, keeping track of scores.
- [FE003](#fe003) User can save the game, at any time, and reload the game from the same saved state.
- [FE004](#fe004) User can change the playing difficulty:
  - [DF002](#df002) Level 1) Simple: Human Click (MVP)
  - [FE005](#fe005) Level 2) Medium: Computer (Random)
  - [FE006](#fe006) Level 3) Hard:Computer AI (Minimax)

---

---

## **2. PLAN**

### _<ins>2.1 User Experience</ins>_

1. _Game Objective_: Explains the objective of the game, which is to place three marks in a row, either horizontally, vertically or diagonally.
2. _Game Play_: Instructuions Describe how the game is played. Each player takes turns placing their mark on an empty cell on the game board. The first player to get three marks in a row wins the game.
3. _Game Board_: Inform that a board which is a 3x3 grid surface. Each cell is empty at the start of the game, and players take turns placing their marks on the board.
4. _User Interaction_: Inform how the user interacts with the game, which is by tapping or clicking on an empty cell to place their mark.
5. _Feedback_: Provide feedback to the user after each move, such as highlighting the winning row or reporting that the game has ended in a draw. Additionally, for invalid moves.
6. _Error Feedback_: Provide feedback to the user when the app or the game logic throw an error from user input or from a runtime exception.
    - Provide graceful degradation in error handling and/or ...
    - Provide graceful degradation when logging console errors and/or ...
    - Give the user an option to either i) choose or ii) contact the developer and log a bug.
7. _Restarting the Game_: Allow the user to easily restart the game, either by clicking a button. <!-- or shaking the device. -->

#### _<ins>2.1.1 Audience(s)</ins>_

- _A) Users_:
    > User experience as players
  - who love board games.
  - who like strategic games.
  - who like tatical games.
  - who like similar games to TicTacToe, a 3,3,3-game, an an [m,n,k-game](https://en.wikipedia.org/wiki/M,n,k-game)
- _B) Junior developers_
   > Developer Experience, as a User Experience
  - who are looking to learn javascript.
  - who are looking to learn object orientated programming.
  - who are looking to learn JDDoc commenting for documentation.
  - who are looking to use text to diagram technologies, like mermaid and techniques like UML etc.

#### _<ins>2.1.2 Requirements</ins>_

1. User Stories

#### _<ins>2.1.3 Roadmap</ins>_

1. Week 1: 2023/03/08 - 2023/03/22: Version 0.1.0 - 0.1.6: Initial Code
2. Week 2: 2023/0/22 - 2023/03/24: Version 0.3.0 - 0.3.4
3. Week 3: 2023/03/21 - 2023/03/23: Version 0.4.0 - 0.4.7

---

---

## **3. DESIGN & UXD**

### _<ins>3.1 App Design</ins>_

#### _3.1.1 UI: HTML &amp; CSS_

The game board is created using HTML div elements styled with CSS.

- Each cell on the board is represented by a div element with a data attributes that stores its state, ident, and row, colium and coordinate position on the board.
- The board is styled using CSS Grid to create a 3x3 grid of cells with inner and outer borders, that creates a faux padding around each cell.
- Each cell as relative border radii.

#### _3.1.2 App: JavaScript_

- The game logic is implemented using JavaScript classes, and files using JS modules, that represent:
  - Class: the _App_, i.e. `__App.js`. Initailises the app on page load, and custom event listeners.
    - UI Helper Class: the _StartListener_, i.e. `__StartListener.js`.
      - _StartListener_:
        - Initialises a new Game instance on each button click
    - UI Helper Class: the _MoveListener_, i.e. `__MoveListener.js`.
      - _MoveListener_:
        - Set up the custom listener, and
        - Listens for each click, and
        - Iniialises the move target cell's data and
        - The current game controler on each move.
  - Class: the _Game_, i.e. `__Game.js`. Game controller
    - Class: the _GamePieces_, i.e. `__GamePieces.js`. Game symbol statemachine, and customised pieces storage, as a future enchancement.
    - Class: the _GameLogic_, i.e. `__GameLogic.js`: Logic machine, randomisation and possibly difficulty levels. as a future enchancement.
    - Class: the _GameBoard_, i.e. `__GameBoard.js`: Statemachine for the game 's grid, and, potentially for  local storage, as a future enchancement.
    - Field: Boolean _this.gameRunning_ is the class level activity state
    - Method: Boolean _this.isAWin()_ returns transitory terminal state of the end of the game
    - Method: Boolean _this.isADraw()_ returns transitory terminal state of the end of the game
  - Utility Classes:
    - Class: the _GameDebug_, a debugger, i.e `__GameDebug.js`: Centralised console debugging statements.
      - Each class & method registers a global console informational logging level for positive flow & output.
      - Each class & method registers a global console error logging level for positive flow & output.

### _<ins>3.2 Diagrams</ins>_

#### _3.2.1 States of the Game_

The state diagram for a game of Tic Tac Toe can be represented with the following states:

- _Start_: The game has not yet started.
<!-- - _Player 1 Choose_: Player 1's choose a symbol.
-   _Player 2 Choose_: Player 2's choose a symbol. -->
- _Player 1 Turn_: Player 1's has a current turn.
- _Player 2 Turn_: Player 2's has a current turn.
- _Player 1 Wins_: Player 1 has won the game.
- _Player 2 Wins_: Player 2 has won the game.
- _Tie Game_: The game has ended in a tie.
<!-- - _Save_: can be activated by any Player to save a game state. -->
- _Start_ can be activated by any Player, and is an alias for _Start_ to start a new game,
  - and can be used a a following state from a terminal state

#### _3.2.2 Transition States of the Game_

Each state can transition to other states based on certain conditions. The possible transitions are:

- _Start_ can transition to _Player 1 Turn_.
- _Player 1 Turn_ can transition to _Player 2 Turn_ or _Player 1 Wins_ or _Tie Game_.
- _Player 2 Turn_ can transition to _Player 1 Turn_ or _Player 2 Wins_ or _Tie Game_.
- _Player 1 Wins_ and _Player 2 Wins_ and _Tie Game_ are terminal states and cannot transition to any other state in the current game.

#### _3.2.3 State Diagram_

```mermaid
stateDiagram-v2
    Start --> Player1Turn
    Player1Turn --> Player2Turn
      Player1Turn --> Player1Wins
      Player1Turn --> TieGame
    Player2Turn --> Player1Turn
      Player2Turn --> Player2Wins
      Player2Turn --> TieGame
    Player1Wins --> End
    Player2Wins --> End
    TieGame --> End
    End --> Restart
    Restart --> NewGame
    NewGame --> Start
    Start: The game has not yet started.
      Player1Turn: It's Player 1's current turn.
      Player2Turn: It's Player 2's current turn.
      Player1Wins: Player 1 has won the game.
      Player2Wins: Player 2 has won the game.
      TieGame: The game has ended in a tie.
    End: Current game terminal state.
    Restart: Resets current game state,
    NewGame: Initialises the next/new game state.
```

#### _3.2.4 User Journey_

> While the state machine of the game, as respresented by the state diagram. i.e. the game flow logic, above represents the phases of the game logic, the user journey focuses on the user experience during the game flow.
> They appear identical, just phrased in different ways and contextes.

In this user journey:

- The game starts with both players choosing their symbols.
- Then, the turns alternate between players
  - The turns alternate until one of them wins or the game ends in a tie.
- At each turn, the player chooses a cell to place their symbol in ...
  - and the game checks for a win or a tie.
- If there is a win, and there is a winner:
  - The game ends and
  - The winning player is declared.
- If the board is full and there is no winner:
  - The game ends in a tie.

```mermaid

graph TD
    A[Start Game] --> B[Player 1 Chooses Symbol]
    B --> C[Player 2 Chooses Symbol]
    C --> D[Player 1 Turn]
    D --> E[Player 1 Chooses Cell]
    E --> F[Check for Win or Tie]
    F --> G[Player 2 Turn]
    G --> H[Player 2 Chooses Cell]
    H --> I[Check for Win or Tie]
    I --> D
    F --> J[End Game: Player 1 Wins]
    I --> K[End Game: Player 2 Wins]
    F --> L[End Game: Tie]
    J ---> M[Display Winning Message]
    K ---> N[Display Winning Message]
    L --> O[Display Tie Message]

```

#### _3.2.5 Code Logic & Activity Flow_

> Activity are similar to both State and User journies, except they focus on the interations between the user and the game logic, i.e. the game methods/functions.

- _Start App_: The app is initialized and the sets ip, for readiness/validity, of UI event listeners (on page load)
  - _Instantiated a New Game_ Create a new game instance
  - _Start StartListener_ Sets up and pass UI & game strings for the start (state) of the game.
  - _Start MoveListener_ Sets up and pass grid UI & game strings after the start (state) of the game.
- _Initialise on Start Button_ Set the start state of the game to truthy, and user click event.
- _Start Game_: A new game is initialized, when game state is true, and game checks for a new game board, game pieces:
  - _Sets the Players's_ To game symbol and pass these to the current game's GamePieces
  - _Initialise Game Pieces_: The game ieces that store each game's pieces and is responsible for switching pieces on each turn.
  - _Initialise Game Board_: The game board is initialsed by clearing the board/setting up the board with empty cells.
  - _Initialise Game Logic_: The game logic is checked and initialised .
- _On Each User's Click_: The onClick checks for
  - _Try Again_ When Grid's cell is occupied, then inform the user to try again.
  - _Available to Move_ When Grid cell is free, the game is actively running, and the end state is not true
    - _Update Grid on Move_: Update the Grid
  - _Ready to Finish_ When the game is not running, and the end state is true
    - _Update Grid at End_: Update the grid/ui and game's state at end of the game.
- _Makes a Move_: Per click, per move, make a move
  - _Checks Valid Move_ Inspects for a valid move
  - _Update Game State_ Sets game active state to true
  - _Update Playing Piece_ Fetch the current playing piece
  - _Check Game Over_ Checks for game terminal state, win or draw or continue.
- _Check for Win or Tie_: The game checks if there is a win or tie on the game board. If true, then ...
  - _End Game?_: The game checks if the game has ended due to a win or tie.
    - _Display Result_: If the game has ended, the game displays the result to the user (either a win or tie) and prompts the user to either restart or exit the game.
        <!-- - _Prompt to Restart_: User is offered to start a new game, or end game.  -->
  - _Restart Game_: If the user chooses to start a new game, then
    - _Reset Game Board_ game resets the game board and starts a new game.
- _Exit Game_: _If the user chooses to exit the game, the game ends._

_Italics_: Optional Functionality: Methods for the game activities. _MVP++_

#### _3.2.6 Class Diagram_

##### _3.2.6.1 App

```mermaid
classDiagram
    class App{
        -GameDebug deBug
        -Game newGame
        -string btnID
        -string moveID
        -string eventTypeid
        -StartListener startListener
        -MoveListener moveListeners
        -string fileName
        -number logLevel
        -number errorLevel
        +App constructor()
        +void onInit()
        +void onError(e)
        +void gameLogger()
    }
```

##### _3.2.6.2 Game Controlers

```mermaid
classDiagram
    class Game {
        -GameDebug deBug
        +string Player1
        +string Player2
        -GamePieces gamePieces(string Player1, string Player2)
        -GameBoard gameBoard
        -GameLogic gameLogic(GameBoard gameBoard, GamePieces gamePieces)
        -boolean gameRunning
        -string fileNamer
        -number logLevel
        -number errorLevel
        +Game constructor( string _X, string O)
        +Game inInit( isRunning)
        +boolean isGameStarted(boolean isGameRun)
        +boolean checkMove(number move, boolean gameState)
        +void updateGrid(number move, Node cell)
        +boolean makeMove(number move, Node current, boolean gameState)
        -void nextTurn(boolean turn, string nextPiece)
        -boolean isGameOver(boolean isRunning)
        -void clearCurrentGame(boolean start)
        -void evaluateParameter(Object param, number argIndex )
    }
```

> Note: Optional param with defaults: log & locname excluded for clarity

##### _3.2.6.3 UI Controlers

```mermaid
classDiagram
   class StartListener{
        -GameDebug debug
        -Node startbutton
        -Game newGame
        -string eventType
        -string btnID
        -string fileName
        -number logLevel
        -number errorLevel
        +StartListener constructor(game,btnID,evnt)
        +void onInit(game)
        +void isStartClicked(): boolean
        +void addListener(element)
        +void onError(error): Error
        +void evaluateParameter(param): Object
        +void evaluateEventType(event,argIndex): string
    }
    class MoveListener{
        -NodeList this.move
        -Game this.game
        -String this.eventType
        -onMove() this.onMove().bind(this)
        -GameDebug this.debug
        +MoveListener constructor(_move, _game, _evt)
        +void setListener()
        +void onMove(i)
    }
```

##### _3.2.6.4 Game Classes/Logic_

```mermaid
classDiagram
    class GameLogic {
        -GameBoard this.gameBoard
        -GamePieces this.gamePieces(_X, _O)
        -GameLogic this.gameLogic(this.gameBoard, this.gamePieces)
        -GameDebug this.debug
        -String this.currentPlayer-Symbol
        -Array[][] this.winningCombinations
        -String this.BLANK
        +GameLogic constructor(GameBoard _board_, GamePieces _pieces)
        +Boolean isAWin()
        +Boolean isAWinCombo()
        +Boolean isADraw()
    }
    class GameBoard {
        -String[] this.grid
        -Boolean this.isValid
        -Boolean this.isInValid
        -GameDebug this.debug
        +GameBoard constructor()
        +Boolean isCellFree()
        +Boolean isCellOccupied()
    }
    class GamePieces {
        -String[] symbol
        -String this.X
        -String this.Y
        -GameDebug this.debug
        +GamePieces constructor(_X, _O)
        +String get X()
        +String get X(piece)
        +String get Y()
        +String set X(piece)
        +String switchPiece(piece)
    }
```

##### _3.2.6.5 Ultility Classes_

```mermaid
classDiagram
    class GameDebug {
        -String this.ERR_HTML
        -String this.ERR_NODE
        +GameDebug constructor()
        +StdErr _debug(obk, mgs, level)
    }
```

#### _3.2.7 Code Flow_


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and adheres to [Semantic Versioning](http://semver.org/).

> - Modified, by incrementing and counting the index of commits on a daily basis
> - (Major.Minor.Patch.CommitIndex) (By CJ)
> - Header Formatting & Intents as below

---

> ## Changelog: v.major.minor.patch.commitindex 20yy.mm.dd
>
> ### `Added`  for new features
>
> ### `Changed` | Updated: for changes in existing functionality
>
> ### `Deprecated`: for soon-to-be removed features
>
> ### `Removed`:  for now removed features, files, folders
>
> ### `Fixed`: for any bug fixes
>
> ### `ToDo`: Project Outstanding items of note
>
> ### `Noted`: Project Logging, Journalling. Comments

---

# Changelog: Project Terni-Lapilli--Toe

> **Template**

## Changelog: 2023.04.10 to 13v.0.5.000.001

### Author Note: Game.js: The Game Controller, and State Machine

See: Too Big for one issue: split between 3 related parts.
[#29: Game.js Part 1: Constrcutor, onIti, isGameStarted, checkMove, updateGrid](https://github.com/iPoetDev/terni-lapilli--toe/issues/29)
[#30: Game.js Part 2: makeMove and sister functions](https://github.com/iPoetDev/terni-lapilli--toe/issues/30)
[#31: Game.js Part 3@ nextTurn, isGameOver, endGame,evaluateParameter](https://github.com/iPoetDev/terni-lapilli--toe/issues/31)

Changelog

- Updated: Bump for all code sections/jsdoc
- Added the new start state constants to this.STATES.game
- See StartListener onInit & ``isStartClicked

- Added Object+Property to constructor,
- Added a state engine to `onInit` , `isGameStarted`, `checkMove`, `makeMove`
- `isGameOver` is better option to pass a predefined state between methods
- Add state engine to yhe class properties. Too late to effect that change of idea.

*State Machine Objects*

- These state engine data structures are excessiive, not DRY
- Messy learning and implementation, for future improvements
- Use a state machine libraries, as problem has been solved by others

- Fixed pieces: this.gamePieces/currenPiece : undefined - I think-
- Todo: Split longer methods into smaller helper methods and enclosures/closure functions.
- Issue: This class has become too big to manually check
- Todo: Need to be split into enums. types, interfaces and other classes.
- Issue: Execessive state management, 100^5 explicit is a symptom no automated test cases
- Problem: State management between classes is hard without testing and tesyt cases
- Design Choice: Class design was a too expensive design approach to fix later on .
- Noted: Game mechanics in IRL use Finite State Machines,
  - Ooops: This project stumbled into this complexity

### Bumped

- `StartListener.js`
  - See <https://api.codestream.com/c/ZBNk4xDOQ2o6RTpe/up6V_D8ATCev4IXWwmdWuw>
- `App.js`
  - <https://github.com/iPoetDev/terni-lapilli--toe/issues/28>

### Added

- `App.js`, `StartListener.js`,
  - Object+Property pattern: Constructor (app object)
  - JSDoc type annotations
  - Local state obejct to onInit, isStartClicked methods -> Further refectoring
    - onInit is bound, so needs to return void, but updates listener startClicked flag
    - isStartClicked returns STATE snd updates the startClicked flag and gameRunning flag

### Changed

- `App.js`
  - Changed: applied the app.property, error.property to method called and param assigns in nominal flow and error flow states
- `StartListener.js`
  - Changed: applied `Listener` object to  constructor, passing this to addListener
  - Updated: addListener using the objects passed property values to improve code readbility.
  - `onError`@ Remodel method, removing the 1st of 2  switches  for a Object+Property pattern, improving readability and logic statement issues.

### No Mutate

- `App.js`
  - No change to `onError` method: @ v0.4.2
  - Remove `onExit`@ v0.4.2
- `StartListener.js`
  - evaluateParameters, evaluateEventType:@ v0.4.1

### Removed

## Changelog: 2023.04.10 to 12v.0.5.000.001

### Author Note

- A Major logic change, a good idea (at 10 days before, i thought),
  - How to solve the problem of multiple states of a game (beyond tic tac toe)
  - The game state: gameRunning: a persistent, mutable, state flag for game activity
  - Ths move state: movesOutcome: a persistent, mutable, state flag for each move's state
  - the results state:  resultsOutcome: a persistent, mutable, state flag for the (no-)results of Win | Draw | Playon
  - How to:
    1) Solve for a game state that has more than binary states i.e. !=== true || false
    2) Solve for a each move's state that has more than binary states i.e. !=== true || false
    3) Solve for a result's state that has more than binary states i.e. !=== true || false
    BOUNUS) Solve for a application's state that has more than binary states i.e.
  - Concern:
    Logic gates based on multiple true || false are indeterminate and multiple are very hard to debug
  - Triggger cause:
    - Code was error free except the key isGameOver was coming back @type string | undefined
    - Upon inspection: the problem space deepened, as there was a mix of multiple booleans, string and other states to evaluated for switch and if conditions.
  -Soltuion approach: Use Strings not Booleans
    - String logic is deterministic, has discrete value and is predicable
    - A Finite state machine, a prototype, was thought off, but not the full ComSci version
  - Outcome:
    - Side Effect: CC scores for methods trippled
    - Side Effect: method lengths exploded (Loc)
    - Side Effect: more methods and helpers to managing state
    - Side Effect: Not using a state management library was not forseen
    - Side Effect: Started with string Contants,ended up with global, class, and method states as data structures.
  - Design Patterns
    - Message & Traces Pattern: Pairing messages and traces for debug to console for eachg step

lib/App.js
// @ts-check
/** App: 001 Module Summary @version 0.5.0 2023/04/13
 *   (c) 2023-2025, Charles Fowler, iPoetDev.github.com
 * @copyright 2023-2025,
 * @file App class for initailising the game conrtoler and input events of TicTacToe
 * @summary: App is the main class of the game, and is responsible for instantiating the game, and the game's listeners.
 * @kind module
 * @export App
 * @type {App}
 * @author Charles Fowler iPoetDev.github.com
 * @date 2023/03/10 @version 0.1.1
 * @since 2023/03/28 @version 0.4.1 @note Issue#21 https://github.com/iPoetDev/terni-lapilli--toe/issues/21
 * @since 2023/03/28 @version 0.4.2 Bump
 * @link  https://github.com/iPoetDev - Profile
 * @link  https://github.com/iPoetDev/terni-lapilli--toe/ - Repository
 * @link  https://github.com/iPoetDev/terni-lapilli--toe/issues/new - Log an Iew Issue
 * @requires { GameDebug } from '__GameDebug.js' @see module:GameDebug
 * @requires { Game } from '__Game.js' @see module:Game
 * @requires { StartListener } from '__StartListener.js' @see module:StartListener
 * @requires { MoveListener } from '__MoveListener.js' @see module:MoveListener
 */
import { GameDebug } from "../lib/__GameDebug.js";
import { Game } from "../lib/__Game.js";
import { StartListener } from "../lib/__StartListener.js";
import { MoveListener } from "../lib/__MoveListener.js";

/** App: 002 Class Definition @version 0.5.0 2023/04/13
 * @name App
 * @kind class
 * @classdesc WebApp interface/initialiser of a game of TicTacToe. Called when page loads, by script tag in index|home.html [ ] TODO Change html.
 * @exports App
 * @type {App}
 * @prop {GameDebug} deBug @member @private
 * @prop {Game} newGame @member @private
 * @prop {StartListener} startListener @member @private
 * @prop {string} butID @member @public
 * @prop {string} moveID @member @public
 * @prop {MoveListener} moveListeners @member @private
 * @prop {string} eventTypeID @member @public
 * @prop {string} fileName @member @private
 * @prop {number} logLevel @member @private
 * @prop {number} errorLevel @member @private
 * @function constructor  Instantiate App  @public @since 0.1.0
//* @function onExit()  Disposed of the event listeners on exit @public  @since 0.4.1
 * @function onError() @throws General or Unknown error @private @since 0.1.5+
 */

/* eslint-disable-next-line require-jsdoc */
/** App: 003 Class Declaration version 0.4.2 @2023/04/04*/
class App {
    /** App: 003 Props/Members @version 0.5.0 @2023/04/13
     * @prop {GameDebug} deBug Class logger @kind member
     * @prop {Game} newGame Current Game instance @kind member
     * @prop {StartListener} startListener Start Game interface @kind member
     * @prop {string} butID Button ID name @kind member @default #start
     * @prop {string} moveID Grid Class name @kind member @default .cell
     * @prop {MoveListener} moveListeners UI Event=Action Interface @kind member
     * @prop {string} eventTypeID Click event type @kind member @default click
     * @prop {string} fileName File's name for tracing @kind member @default App.js
     * @prop {number} logLevel Class logging level for info console @kind member @default 0
     * @prop {number} errorLevel Class logging level for error console @kind member @default 0
     * @version 0.4.0 @date 2023/03/28  Enabled customListeners, updated Class props
     * @version 0.4.1 @date 2023/04/01.  @note GitHub.Issue#21/#18/#11 Updated constructor's logging and tracing, removed logging methods
     * @version 0.4.2 @date 2023/04/03 Bumped, JSDoc tidy
     * @link https://github.com/iPoetDev/terni-lapilli--toe/issues/21
     */
    // App Class Members
    /**@type {GameDebug} */
    deBug;
    /**@type {Game} */
    newGame = new Game();
    /**@type {string} */
    fileName = "App.js";
    /**@type {number} */
    logLevel = 1; /**@since 0.4.0 */
    /**@type {number} */
    errorLevel = 9; /**@since 0.4.0 */

    /** App: 003 Class Constructor @version 0.5.0 @date 2023/04/04
     * @function constructor
     * @kind function
     * @classdec Create an instance of running App.
     * @constructs App
     * @param {string} [locname] @default `App,js`
     * @type {App}
     * @memberof App
     * @date 2023/03/22 @version 0.3.0
     * @date 2023/03/24. @version 0.4.0 @note GitHub.Issue#3 @link https://github.com/iPoetDev/terni-lapilli--toe/issues/3
     * @date 2023/03/31. @version 0.4.1
     * @date 2023/04/03. @version 0.4.2 Message+Trace pattern, Improved readability, added @ts-ignore for code hotspots
     * @date 2023/04/03. @version 0.5.0 Bumped: Object+Property Pattern
     */
    constructor(locname = `App.js`) {
        // 1: Instance's global Debug and Console log
        this.deBug = new GameDebug();
        const app = {
            player1: "X",
            player2: "Y",
            id: { start: "#start", moves: ".cell", event: "click", },
            /** @type {Game} */
            game: this.newGame,
            /**@type {StartListener|undefined} */
            button: undefined,
            /**@type {MoveListener|undefined} */
            grid: undefined,
            traces: {
                location: this.fileName,
                game: `${locname}: new Game`,
                start: `${locname}: new StartListener`,
                move: `${locname}: new MoveListener`,
            },
            log: {
                /**@type {number} */
                console: 1,
                /**@type {number} */
                stderr: 9,
            },
        };
        // 2: Set new Game instance, else throw new Game Error
        try {
            app.game = new Game(app.player1, app.player2, app.log.console, app.traces.game);
        } catch (constructorerror) {
            this.onError(constructorerror, 1, app.log.stderr, app.traces.game);
        }
        // 3: Set new StartListener instance, else throw new StartListener Error
        try {
            // this.startListener = new StartListener(
            app.button = new StartListener(
                app.game,
                //this.startID,
                app.id.start,
                app.id.event,
                app.log.console,
                app.traces.start
            );
        } catch (starterror) {
            const errors = {
                error: starterror,
                flag: 1,
                traces: {
                    line: `startListeners ${starterror.linenums}`,
                    trace: `${app.traces.location}: Instance of startListener`,
                    error: `${app.traces.start}, id: ${app.id.start} button, on a ${app.id.event}`,
                },
                console: app.id.stderr,
            };
            this.deBug._debug(errors.error, errors.traces.line, errors.console, errors.traces.trace);
            this.onError(errors.error, errors.flag, errors.console, errors.traces.error);
        }
        //4: Set new MoveListener instanxe, else throw new StartListener Error
        try {
            app.grid = new MoveListener(
                // @ts-ignore
                app.game,
                app.id.moves,
                app.id.event,
                app.log.console,
                app.traces.move
            );
        } catch (moveerror) {
            const errors = {
                error: moveerror,
                flag: 1,
                traces: {
                    line: `moveListeners ${moveerror.linenums}`,
                    trace: `${app.traces.location}: Instance of startListener`,
                    error: `${app.traces.move}, id: ${app.id.moves} div/target, on a ${app.id.event}`,
                },
                console: app.id.stderr,
            };
            this.deBug._debug(errors.error, errors.traces.line, errors.console, errors.traces.trace);
            this.onError(errors.error, errors.flag, errors.console, errors.traces.error);
        }
    }

    /** onExit @todo [ ] TODO ENHANCEMENT
     * @function onExit
     * @kind function
     * @description This function handles the exit state of the application class, Runtime exit function for the Application level
     * @date 2023/03/31 @version 0.4.1 @note Issue#21 @link [#21]https://github.com/iPoetDev/terni-lapilli--toe/issues/21
    onExit() {
        // Clean up and remove all listeners
        // A good coding practice?
        this.startListener.removeListener();
        // this.moveListeners.removeListeners();
    }

    /** onError @version 0.4.2. @date 2023/04/04
     * @function onError
     * @kind function
     * @description This function handles the error state of the application class, Runtime error function for the Application level error handling
     * @usage Call when the app is initaitaied and when is initially run.
     * @param {Error} error General, {*} error object
     * @param {number} errflag 1: General Error, Default: Unknown Error. @default 0 : Unknown
     * @param {number} log Debugging error level: Full Stack Inspect @default this.errorLevel
     * @param {string} locname Filename locator for onError/Console @default App.js
     * @throws {Error} Application error, logs to console.
     * @throws {UncaughtError} Uncaught Errors
     * @memberof App
     * @version 0.1.0 @date 2023/03/15
     * @version 0.3.0 @date 2023/03/20
     * @version 0.4.0 @date 2023/03/24  .GitHub [#8](https://github.com/iPoetDev/terni-lapilli--toe/issues/8)
     * @version 0.4.1 @date 2023/03/28 Adjusted logging and tracing to console on errors
     * @version 0.4.2 @date 2023/04/04 Bump. JSDoc tidy, ts-ignore for code hotspots/TS issues, add block Constant, Message&Trace
     * @note Issue#21 [#21](https://github.com/iPoetDev/terni-lapilli--toe/issues/21)
     */
    //  Make this more robust, covering more edge cases and handling errors:
    onError(error, errflag = 0, log = this.errorLevel, locname = `${this.fileName}: onError`) {
        const check = log;
        const GENERAL = 1;
        try {
            switch (errflag) {
                case GENERAL:
                    if (error) {
                        const generalmessage = `${locname}: General Error ${errflag}: App Init Error: ${error.name} - ${error.stack}`;
                        const generaltrace = `${locname}: General Error ${error.name}`;
                        // General Error & Debugging debug message.
                        this.deBug._debug(error, generalmessage, check, generaltrace);
                        throw new Error(generalmessage);
                    }
                    break;
                default: //  Unknown/Uncaught Exceptions:
                    const unknownmessage = `${locname}: Unknown Error ${errflag}: Application Error: ${error.name} - ${error.stack}`;
                    const unknowntrace = `${locname}: Unknown Issue: onError()`;
                    this.deBug._debug(error, unknownmessage, check, unknowntrace);
                    break;
            }
        } catch (uncaughterror) {
            const uncaughtmessage = `${locname}: Unknown Error ${errflag}: Application Error: ${uncaughterror.name} - ${uncaughterror.stack}`;
            const uncaughttrace = `${locname}: Uncaught Issue: onError()`;
            this.deBug._debug(uncaughterror, uncaughtmessage, check, uncaughttrace);
        }
    }
}

export { App };
